connection-cache/src/connection_cache.rs

```diff
@@ -102,48 +154,37 @@ where
         // Read again, as it is possible that between read lock dropped and the write lock acquired
         // another thread could have setup the connection.
 
-        let should_create_connection = map
+        let (need_new_connection, should_create_connection) = map
             .get(addr)
             .map(|pool| pool.need_new_connection(self.connection_pool_size))
-            .unwrap_or(true);
+            .unwrap_or((true, true));
 
         let (cache_hit, num_evictions, eviction_timing_ms) = if should_create_connection {
-            // evict a connection if the cache is reaching upper bounds
-            let mut num_evictions = 0;
-            let mut get_connection_cache_eviction_measure =
-                Measure::start("get_connection_cache_eviction_measure");
-            let existing_index = map.get_index_of(addr);
-            while map.len() >= MAX_CONNECTIONS {
-                let mut rng = thread_rng();
-                let n = rng.gen_range(0..MAX_CONNECTIONS);
-                if let Some(index) = existing_index {
-                    if n == index {
-                        continue;
-                    }
-                }
-                map.swap_remove_index(n);
-                num_evictions += 1;
-            }
-            get_connection_cache_eviction_measure.stop();
-
-            map.entry(*addr)
-                .and_modify(|pool| {
-                    pool.add_connection(&self.connection_config, addr);
-                })
-                .or_insert_with(|| {
-                    let mut pool = self.connection_manager.new_connection_pool();
-                    pool.add_connection(&self.connection_config, addr);
-                    pool
-                });
-            (
-                false,
-                num_evictions,
-                get_connection_cache_eviction_measure.as_ms(),
+            Self::create_connection_internal(
+                self.connection_config.clone(),
+                self.connection_manager.clone(),
+                &mut map,
+                addr,
+                self.connection_pool_size,
+                None,
             )
         } else {
             (true, 0, 0)
         };
 
+        if need_new_connection && !should_create_connection {
```

Comment:
This logic is pretty confusing... can you please clarify in a comment that we want to "warm up" the connection pool the first time it's used (I assume)?